### PR TITLE
[Foundation] Scale back struct Notification's userInfo to be in-line …

### PR DESCRIPTION
**Pull request by phausler**
#### What's in this pull request?

This simplifies the bridging story for Notifications to their objc counterparts since the id -> Any and AnyHashable changes have now been applied (which makes the previous boxing strategy no longer needed). Previous consumers of Notification that were using String keys should still work, however any explicit dictionary types should migrate from Swift 2.2 -> Swift 3 from userInfo as [NSObject:AnyObject] to [AnyHashable:Any]. The condition of distributed notifications (in non sandboxed apps) requiring plist types still applies and will fail at runtime if incorrect types are passed into the objective-c layer, and in the case of sandboxed apps userInfo still is forbidden (this change is a non functional change in the respect to those behaviors).
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

rdar://problem/27426757
rdar://problem/27561621
rdar://problem/27259984

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci Please test OS X platform

File: test/1_stdlib/TestUserInfo.swift [hunk 1]
```diff
@@ -85,20 +85,27 @@ class TestUserInfo : TestUserInfoSuper {
...
+            AnyHashable(userInfoKey) : "this can convert"
+        ]
+        let info3: [AnyHashable : Any] = [
+            AnyHashable(userInfoKey) : testStructure2
         ]
 
         let note1 = Notification(name: notifName, userInfo: info1)
         let note2 = Notification(name: notifName, userInfo: info1)
-        expectNotEqual(note1, note2)
+        expectEqual(note1, note2)
```
Comment (phausler) [inline test/1_stdlib/TestUserInfo.swift]: notifications created with the same name and same userInfo should be equal

File: test/1_stdlib/TestUserInfo.swift [hunk 2]
```diff
@@ -85,20 +85,27 @@ class TestUserInfo : TestUserInfoSuper {
...
 
         let note3 = Notification(name: notifName, userInfo: info2)
         let note4 = Notification(name: notifName, userInfo: info2)
         expectEqual(note3, note4)
+
+        let note5 = Notification(name: notifName, userInfo: info3)
+        expectNotEqual(note1, note5)
```
Comment (phausler) [inline test/1_stdlib/TestUserInfo.swift]: differing userInfo should not be equal

File: test/1_stdlib/TestUserInfo.swift [hunk 3]
```diff
@@ -59,8 +59,8 @@ class TestUserInfo : TestUserInfoSuper {
         let userInfoKey = "userInfoKey"
         let notifName = Notification.Name(rawValue: "TestSwiftNotification")
         let testStructure = SomeStructure(i: 5, str: "10", sub: SubStruct(i: 6, str: "11"))
-        let info: [String : Any] = [
-            userInfoKey : testStructure
+        let info: [AnyHashable : Any] = [
+            AnyHashable(userInfoKey) : testStructure
```
Comment (michael-lehew) [inline test/1_stdlib/TestUserInfo.swift]: this is unfortunate, but fine.

Comment (phausler) [inline test/1_stdlib/TestUserInfo.swift]: Yea it is fallout from the AnyHashable being a structure and not a protocol (and not being transparently convertible to the counterpart

**Comment by michael-lehew** [comment]
changes look good to me :)

**Comment by lattner** [comment]
@phausler and @michael-lehew : FYI, @rjmccall is working on resolving the missing implicit conversion to AnyHashable issue, for Swift 3.

**Comment by lattner** [comment]
This is the right thing to do in the meantime though.

**Comment by tkremenek** [comment]
@swift-ci test

**Pull request merged**